### PR TITLE
Issue #22057: Add search term groups telemetry

### PR DESCRIPTION
**Pull request by rocketsroger**
### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by rocketsroger** [comment]
# Request for data collection review form

All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.

1. What questions will you answer with this data?
  * This data will help us understand how users are interacting with search term groups feature.
2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? 
  * This will help us understand if search term groups feature is useful to the user. 
3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
  * There are no other alternatives. 
4. Can current instrumentation answer these questions?
  * No. 
5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki. 
Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.

Measurement Description | Data Collection Category | Tracking Bug #
-- | -- | --
Search term group enabled: If the search term group feature is enabled by the user.	| Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/22057
Number of search term groups: how many groups does the user have when they tap on the tabs tray	| Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/22057
Average tabs per search term group: how many tabs per group on average does the user have when they tap on the tabs tray	| Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/22057
Jump back in group tapped: The user tapped on a jump back in search term group | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/22057

6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
  * https://dictionary.telemetry.mozilla.org/apps/fenix 
7. How long will this data be collected? Choose one of the following:
  * One year (ending on November 1st, 2022), with the option to renew at that point. 
8. What populations will you measure?
  * All channels, all locales, all countries 
9. If this data collection is default on, what is the opt-out mechanism for users?
  * Default Glean SDK opt-out mechanism. 
10. Please provide a general description of how you will analyze this data.
  * Glean and Amplitude. 
11. Where do you intend to share the results of your analysis?
  *  Only on Glean, Amplitude, and with mobile teams. 
12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
   * No third-party tools.

File: app/metrics.yaml [hunk 1]
```diff
@@ -1667,6 +1667,19 @@ preferences:
     notification_emails:
       - android-probes@mozilla.com
     expires: "2022-11-01"
+  search_term_groups_enabled:
```
Comment (eliserichards) [inline app/metrics.yaml:1670]: Is this tracking if the feature is ON, or if the feature is SEEN by the user? (I can't remember what we defined for this)

**Comment by eliserichards** [comment]
> > # Request for data collection review form
> 
> All questions are mandatory. You must receive review from a data steward peer on your responses to these questions before shipping new data collection.
> 
> 1. What questions will you answer with this data?
>   * This data will help us understand how users are interacting with search term groups feature.
> 2. Why does Mozilla need to answer these questions? Are there benefits for users? Do we need this information to address product or business requirements? 
>   * This will help us understand if search term groups feature is useful to the user. 
> 3. What alternative methods did you consider to answer these questions? Why were they not sufficient?
>   * There are no other alternatives. 
> 4. Can current instrumentation answer these questions?
>   * No. 
> 5. List all proposed measurements and indicate the category of data collection for each measurement, using the Firefox data collection categories found on the Mozilla wiki. 
> Note that the data steward reviewing your request will characterize your data collection based on the highest (and most sensitive) category.
> 
> Measurement Description | Data Collection Category | Tracking Bug #
> -- | -- | --
> Search term group enabled: If the search term group feature is enabled by the user.	| Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/22057
> Number of search term groups: how many groups does the user have when they tap on the tabs tray	| Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/22057
> Average tabs per search term group: how many tabs per group on average does the user have when they tap on the tabs tray	| Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/22057
> Jump back in group tapped: The user tapped on a jump back in search term group | Category 2 - interaction data | https://github.com/mozilla-mobile/fenix/issues/22057
> 
> 6. Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.
>   * https://dictionary.telemetry.mozilla.org/apps/fenix 
> 7. How long will this data be collected? Choose one of the following:
>   * One year (ending on November 1st, 2022), with the option to renew at that point. 
> 8. What populations will you measure?
>   * All channels, all locales, all countries 
> 9. If this data collection is default on, what is the opt-out mechanism for users?
>   * Default Glean SDK opt-out mechanism. 
> 10. Please provide a general description of how you will analyze this data.
>   * Glean and Amplitude. 
> 11. Where do you intend to share the results of your analysis?
>   *  Only on Glean, Amplitude, and with mobile teams. 
> 12. Is there a third-party tool (i.e. not Telemetry) that you are proposing to use for this data collection?
>    * No third-party tools. 
> 

# Data Review Form (to be filled by Data Stewards)

Instructions: Data Stewards will review a request for data collection and endorse responses to each question. **If the request does not provide answers to questions, reviewers give an r- and point to the questions that can’t be answered.**

1) Is there or will there be **documentation** that describes the schema for the ultimate data set in a public, complete, and accurate way? Click the documentation link provided in Q6 and ensure it is publicly accessible and does or will contain documentation for the data collection. (see [here](https://github.com/mozilla/activity-stream/blob/master/docs/v2-system-addon/data_dictionary.md), [here](https://github.com/mozilla-mobile/focus/wiki/Install-and-event-tracking-with-the-Adjust-SDK), and [here](https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/telemetry/index.html) for examples).  Refer to the appendix for "documentation" if more detail about documentation standards is needed.

     * Yes, through the metrics.yaml and the [Glean Dictionary](https://dictionary.telemetry.mozilla.org/)

2) Is there a control mechanism that allows the user to turn the data collection on and off? (Note, for data collection not needed for security purposes, Mozilla provides such a control mechanism) Provide details as to the control mechanism available.

     * Yes, through the "Send Usage Data" preference in the application settings
     
3) If the request is for permanent data collection, is there someone who will monitor the data over time?

     * N/A, the data will be renewed or removed by November 1, 2022.

4) Using the **[category system of data types](https://wiki.mozilla.org/Data_Collection)** on the Mozilla wiki, what collection type of data do the requested measurements fall under?

     * Category 2 - interaction data

5) Is the data collection request for default-on or default-off?

     * Default-on

6) Does the instrumentation include the addition of **any *new* identifiers** (whether anonymous or otherwise; e.g., username, random IDs, etc.  See the appendix for more details)?

     * No

7) Is the data collection covered by the existing Firefox privacy notice? 

     * Yes

8) Does the data collection use a third-party collection tool? 

     * No


## Result

data-review+

File: app/metrics.yaml [hunk 2]
```diff
@@ -1667,6 +1667,19 @@ preferences:
...
+    type: boolean
+    description: |
+      Is search term group in tabs tray on?
+    bugs:
+      - https://github.com/mozilla-mobile/fenix/issues/22057
+    data_reviews:
+      - https://github.com/mozilla-mobile/fenix/pull/22058
```
Comment (eliserichards) [inline app/metrics.yaml:1677]: Don't forget to update these data review links 👍

File: app/metrics.yaml (hunk 1 above)
Comment (rocketsroger) [reply app/metrics.yaml:1670]: This is just to track if the feature is on.

File: app/src/main/java/org/mozilla/fenix/tabstray/browser/TabSorter.kt [hunk 3]
```diff
@@ -48,6 +48,9 @@ class TabSorter(
         // We don't need to provide a selectedId, because the [TabGroupAdapter] has that built-in with support from
         //  NormalBrowserPageViewHolder.scrollToTab.
         val (groups, remainderTabs) = searchTermTabs.toSearchGroups(groupsSet)
+        val averageTabsPerGroup = groups.map { it.tabs.size }.average().let { if (it.isNaN()) 0.0 else it }
```
[the remark this answers is not in the document]
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/tabstray/browser/TabSorter.kt]: After some thoughts, you're right that we should just avoid reporting this value when there's no search term group(s).

**Pull request merged**